engines/scumm/gfx_mac.cpp

```diff
@@ -119,235 +133,3824 @@ void ScummEngine::mac_drawStripToScreen(VirtScreen *vs, int top, int x, int y, i
+	pt.clear();
+	clearClickedStatus();
+
+	_messageBannerActive = false;
+
+	return ks;
+}
+
+// ===========================================================================
+// Macintosh GUI
+// ===========================================================================
+
+// ---------------------------------------------------------------------------
+// Simple window and widget classes. These are meant to emulate the look and
+// feel of an 80s Macintosh.
+// ---------------------------------------------------------------------------
+
+// ---------------------------------------------------------------------------
+// Base widget
+// ---------------------------------------------------------------------------
+
+MacGui::MacWidget::MacWidget(MacGui::MacDialogWindow *window, Common::Rect bounds, Common::String text, bool enabled) : MacGuiObject(bounds, enabled), _window(window), _text(text) {
+	// Widgets are clipped to the inner surface of the dialog. If a widget
+	// is clipped out of existence, make it invisible to avoid crashes.
+
+	Graphics::Surface *s = _window->innerSurface();
+
+	_bounds.clip(Common::Rect(s->w, s->h));
+
+	if (_bounds.width() <= 0 || _bounds.height() <= 0)
+		_visible = false;
+}
+
+bool MacGui::MacWidget::findWidget(int x, int y) const {
+	return _enabled && _bounds.contains(x, y);
+}
+
+void MacGui::MacWidget::setRedraw(bool fullRedraw) {
+	if (fullRedraw)
+		_fullRedraw = true;
+	else
+		_redraw = true;
+}
+
+void MacGui::MacWidget::setEnabled(bool enabled) {
+	_enabled = enabled;
+	setRedraw(true);
+}
 
-	int x = 216;
-	int y = 377;
-	int width = 62;
-	int height = 22;
-	int var = 50;
+void MacGui::MacWidget::setValue(int value) {
+	_value = value;
+	setRedraw();
+}
 
-	byte *ptr = (byte *)_macScreen->getBasePtr(x,  y);
-	int pitch = _macScreen->pitch;
+void MacGui::MacWidget::drawBitmap(Common::Rect r, const uint16 *bitmap, Color color) const {
+	_window->_gui->drawBitmap(_window->innerSurface(), r, bitmap, color);
+}
 
-	int slot = getVerbSlot(53, 0);
-	VerbSlot *vs = &_verbs[slot];
+int MacGui::MacWidget::drawText(Common::String text, int x, int y, int w, Color fg, Color bg, Graphics::TextAlign align, bool wordWrap, int deltax) const {
+	if (text.empty())
+		return 0;
 
-	vs->curmode = (VAR(var) != 0);
-	vs->curRect.left = x / 2;
-	vs->curRect.right = (x + width) / 2;
-	vs->curRect.top = y / 22;
-	vs->curRect.bottom = (y + height) / 2;
+	const Graphics::Font *font = _window->_gui->getFont(kSystemFont);
 
-	_macScreen->fillRect(Common::Rect(x, y, x + width, y + height), 0);
+	// Apply text substitutions
 
-	if (VAR(var)) {
-		for (int w = 1; w < width - 1; w++) {
-			ptr[w] = 7;
-			ptr[w + pitch * (height - 1)] = 7;
+	for (uint i = 0; i < text.size() - 1; i++) {
+		if (text[i] == '^') {
+			uint nr = text[i + 1] - '0';
+			if (_window->hasSubstitution(nr)) {
+				Common::String &subst = _window->getSubstitution(nr);
+				text.replace(i, 2, subst);
+			}
 		}
+	}
+
+	// Word-wrap text
```

Comment:
The algo suspiciously similar to `Font::wordWrapText()`, but perhaps you have your own for better control